assets/src/dashboard/components/popoverMenu/test/story-menu-generator.js

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/**
+ * Internal dependencies
+ */
+import { noop } from '../../../../design-system';
+import {
+  STORY_CONTEXT_MENU_ITEMS,
+  STORY_CONTEXT_MENU_ACTIONS,
+} from '../../../constants';
+import { generateStoryMenu } from '../story-menu-generator';
+
+const TEST_MENU_ITEM_ACTIONS = Object.values(STORY_CONTEXT_MENU_ACTIONS).reduce(
+  (all, actionName) => {
+    all[actionName] = noop;
+
+    return all;
+  },
+  {
+    default: noop,
+  }
+);
+
+const TEST_STORY = {
+  previewLink: 'preview-link',
+  bottomTargetAction: 'bottom-target-action',
+};
+
+describe('generateStoryMenu', () => {
+  it('should generate menu items of the correct shape', () => {
+    const menuItems = generateStoryMenu({
+      menuItemActions: TEST_MENU_ITEM_ACTIONS,
+      menuItems: STORY_CONTEXT_MENU_ITEMS,
+      story: TEST_STORY,
+    });
+
+    expect(menuItems).toStrictEqual([
+      expect.objectContaining({
+        href: 'bottom-target-action',
+        label: 'Open in editor',
+        newTab: false,
+        onClick: expect.any(Function),
```

Comment:
The tests aren't testing that the correct functions were assigned to the correct list items. The `value` is used to associate the event handler with the list item, but the `value` is not returned on the final list item so it's tricky to identify which ones are which.

We could use the labels, but any update to the labels will break this test. I think that wouldn't be a good thing.

This is another indication that there must be a simpler way to create the menu items with the correct props and test them easily.

Reply:
We could always click on them and test that they output a string (like the value of the item) though. That would make these tests better

Reply:
It's ultimately a matter of preference for how we'd like the storyMenu props API. With the current impl, I think this test is proper.

I think the Actions object is an improvement to the `storyMenu` API because the data for the `menuItem` is closer together now (i.e. being provided in the same prop rather than `onMenuItemSelected`). There is a trade-off about this solution... what was explicit about the switch statement in the `handleMenuItemSelected` functions is the nature of the `value` property being a key to relate the function to the item and `generate` just hydrated some special props. You've offloaded that to `generate` and so you want a test to assure that the actions are properly related to the items using the `value` property. But, since if a key is not available in the Actions object, you provide a noop, it should _always_ provide a function. That's tested here.

Could the `menuItem`s could provide their own `onClick` or `onSelect` where they are memoized by mapping over them and using their `value` to attach the correct `onClick`? then the props just get spread by `generate`, and you could test that `generate` is providing whatever you attach to the `menuItem` itself in this test. The trade off here is that makes validating the `menuItem` shape/props a little harder.

Reply:
I like this idea a lot @embarks. I think it'll be a lot easier to read if we don't have the actions being drilled down alongside the menu items. Setting it up at the beginning is much better.

Reply:
I was initially trying to set these up in the `storiesView.js`:

```
const enabledMenuItems = useMemo(() => {
    const items = enableInProgressStoryActions
      ? STORY_CONTEXT_MENU_ITEMS
      : STORY_CONTEXT_MENU_ITEMS.filter((item) => !item.inProgress);

    return items.map((menuItem) => {
      let onClick = () => setContextMenuId(-1);

      switch (menuItem.value) {
        case STORY_CONTEXT_MENU_ACTIONS.COPY_STORY_LINK:
          onClick = handleCopyStoryLink;
          break;
        case STORY_CONTEXT_MENU_ACTIONS.CREATE_TEMPLATE:
          onClick = handleCreateTemplateFromStory;
          break;
        case STORY_CONTEXT_MENU_ACTIONS.DELETE:
          onClick = handleDeleteStory;
          break;
        case STORY_CONTEXT_MENU_ACTIONS.DUPLICATE:
          onClick = handleDuplicateStory;
          break;
        case STORY_CONTEXT_MENU_ACTIONS.OPEN_STORY_LINK:
          onClick = handleOpenStoryInEditor;
          break;
        case STORY_CONTEXT_MENU_ACTIONS.RENAME:
          onClick = handleRenameStory;
          break;
        default:
          break;
      }

      return { ...menuItem, onClick };
    });
  }, [
    enableInProgressStoryActions,
    handleCopyStoryLink,
    handleCreateTemplateFromStory,
    handleDeleteStory,
    handleDuplicateStory,
    handleOpenStoryInEditor,
    handleRenameStory,
  ]);
```

I couldn't get this to work though 😕. I was running into problems because these functions receive a `story` as an argument, but the way we were doing that was setting that up in `generateStoryMenu` 😕. I tried to curry these to account for the stories, but I ran into problems with rendering. I was getting a react bug that said that it couldn't render `<StoryMenu />` while `<StoryView />`  was rendering. I tried tinkering with it, but after a while gave up and decided to leave it.

imo it would be better to create these event handlers in a component being rendered in the `stories.map` (like the `<StoryMenu />`) since each one would be able to access a specific story. However, we'd have to prop drill all the stuff that the event handlers would need (`storyActions`, `setContextMenuId`, `setRenameId`, etc).

Don't love how I currently have it set up but I think I may keep it like this for now. The current api isn't super easy to work with 😭

Reply:
I agree, it's still an improvement and the explanation and code here is great documentation for revisiting it in the future.